astropy/utils/tests/test_data.py

```diff
@@ -430,62 +904,456 @@ def __init__(self, data):
+    with open(cd, "w") as f:
+        f.write(ct)
+    with paths.set_temp_cache(tmpdir):
+        check_quietly_ignores_bogus_cache()
+    assert get_file_contents(cd) == ct
+    os.remove(cd)
+
+    os.makedirs(cd)
+    py_version = "py" + str(sys.version_info.major)
+    cd = str(tmpdir / "astropy" / "download" / py_version)
+    with open(cd, "w") as f:
+        f.write(ct)
+    with paths.set_temp_cache(tmpdir):
+        check_quietly_ignores_bogus_cache()
+    assert get_file_contents(cd) == ct
+    os.remove(cd)
+
+    cd = str(tmpdir / "astropy" / "download" / py_version / "urlmap")
+    os.makedirs(cd)
+    with paths.set_temp_cache(tmpdir):
+        check_quietly_ignores_bogus_cache()
+
+
+def test_get_fileobj_str(a_file):
+    fn, c = a_file
+    with get_readable_fileobj(str(fn)) as rf:
+        assert rf.read() == c
+
+
+def test_get_fileobj_localpath(a_file):
+    fn, c = a_file
+    with get_readable_fileobj(py.path.local(fn)) as rf:
+        assert rf.read() == c
+
+
+def test_get_fileobj_pathlib(a_file):
+    fn, c = a_file
+    with get_readable_fileobj(pathlib.Path(fn)) as rf:
+        assert rf.read() == c
+
+
+def test_get_fileobj_binary(a_binary_file):
+    fn, c = a_binary_file
+    with get_readable_fileobj(fn, encoding="binary") as rf:
+        assert rf.read() == c
+
+
+def test_get_fileobj_already_open_text(a_file):
+    fn, c = a_file
+    with open(fn, "r") as f:
+        with get_readable_fileobj(f) as rf:
+            with pytest.raises(TypeError):
+                rf.read()
+
+
+def test_get_fileobj_already_open_binary(a_file):
+    fn, c = a_file
+    with open(fn, "rb") as f:
+        with get_readable_fileobj(f) as rf:
+            assert rf.read() == c
+
+
+def test_get_fileobj_binary_already_open_binary(a_binary_file):
+    fn, c = a_binary_file
+    with open(fn, "rb") as f:
+        with get_readable_fileobj(f, encoding="binary") as rf:
+            assert rf.read() == c
+
+
+def test_cache_contents_not_writable(temp_cache, valid_urls):
+    c = cache_contents()
+    with pytest.raises(TypeError):
+        c["foo"] = 7
+    u, _ = next(valid_urls)
+    download_file(u, cache=True)
+    c = cache_contents()
+    assert u in c
+    with pytest.raises(TypeError):
+        c["foo"] = 7
+
+
+def test_cache_read_not_writable(temp_cache, valid_urls):
+    with _cache() as (dldir, urlmap):
+        with pytest.raises(TypeError):
+            urlmap["foo"] = 7
+    u, _ = next(valid_urls)
+    download_file(u, cache=True)
+    with _cache() as (dldir, urlmap):
+        assert u in urlmap
+        with pytest.raises(TypeError):
+            urlmap["foo"] = 7
+
+
+def test_cache_not_relocatable(tmpdir, valid_urls):
+    u, c = next(valid_urls)
+    d1 = tmpdir / "1"
+    os.mkdir(d1)
+    with paths.set_temp_cache(d1):
+        p1 = download_file(u, cache=True)
+        assert is_url_in_cache(u)
+        assert get_file_contents(p1) == c
+    d2 = tmpdir / "2"
+    # this will not work!
```

Comment:
I don't quite understand this comment. It says "this will not work" but no error is being caught. Can you please elaborate?